Disable pattern matching in val syntax in sbt 0.13.7.

jsuereth opened the pull request:
Fixes #1661 or at least works around it for now.

Review by @eed3si9n cc> @jaceklaskowski

jsuereth committed:
Disable pattern matching in val syntax in sbt 0.13.7.

Fixes #1661 or at least works around it for now.

eed3si9n commented on main/src/main/scala/sbt/internals/parser/SbtParser.scala:
```diff
@@ -92,6 +92,21 @@ private[sbt] case class SbtParser(file: File, lines: Seq[String]) extends Parsed
+      t match {
+        case (x @ (toolbox.u.ValDef(_, _, _, _) | toolbox.u.DefDef(_, _, _, _, _, _))) =>
+          val content = modifiedContent.substring(x.pos.start, x.pos.end)
+          val prettyPrint = x.toString
```
@xuwei-k said isn't there a better API than using toString https://twitter.com/xuwei_k/status/520639740274417664. We could check that the last parameter in `ValDef` and `DefDef` are not `EmptyTree` and get the same effect as checking for `=`.

eed3si9n commented:
Travis is on a roll today.

```
Running run / concurrent
[info] Getting org.scala-sbt sbt 0.13.7-SNAPSHOT ...
[info] :: retrieving :: org.scala-sbt#boot-app
[info]  confs: [default]
[info]  51 artifacts copied, 0 already retrieved (15632kB/93ms)
[info] Getting Scala 2.10.4 (for sbt)...
[info] :: retrieving :: org.scala-sbt#boot-scala
[info]  confs: [default]
[info]  5 artifacts copied, 0 already retrieved (24459kB/67ms)
[info] [error] [/tmp/sbt_1bdda655/concurrent/build.sbt]:35: Pattern matching in val statements is not supported
[info] Project loading failed: (r)etry, (q)uit, (l)ast, or (i)gnore? 
[error] x run / concurrent
```

Looking at https://github.com/sbt/sbt/blob/ddf8a173b0a0bdb8f3f5f129701ec9eba429a5c1/sbt/src/sbt-test/run/concurrent/build.sbt, I'm guessing that it's tripping on:

``` scala
def waitFor(f: File) {
    if(!f.exists) {
        Thread.sleep(300)
        waitFor(f)
    }
}
```

This in fact does _not_ have `=`, but likely the pretty print of Tree would desugar this to

``` scala
def waitFor(...): Unit = { .... }
```

eed3si9n commented on main/src/main/scala/sbt/internals/parser/SbtParser.scala:
```diff
@@ -92,6 +92,21 @@ private[sbt] case class SbtParser(file: File, lines: Seq[String]) extends Parsed
+    // Check No val (a,b) = foo *or* val a,b = foo as these are problematic to range positions and the WHOLE architecture.
+    def isBadValDef(t: Tree): Boolean =
+      t match {
+        case (x @ (toolbox.u.ValDef(_, _, _, _) | toolbox.u.DefDef(_, _, _, _, _, _))) =>
```
Given the procedure syntax (`def foo {}`), we should just match on `ValDef`, and not `DefDef`.

eed3si9n committed:
Remove DefDef and check rhs

- lhs pattern matching occurs only for vals.
- def also breaks due to proc notation.
- instead of pretty printing, we can just check rhs.

jsuereth commented:
@eed3si9n - I just rebased.  Any comments on this pr?

eed3si9n commented:
LGTM

eed3si9n commented:
Manually merged this.
